Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -132,17 +134,92 @@ public async Task Push(Layer layer, string name, Action<string> logProgressMessa
             x = new UriBuilder(new Uri(BaseUri, pushResponse.Headers.Location?.OriginalString ?? ""));
         }
 
+        Uri patchUri = x.Uri;
+
+        x.Query += $"&digest={Uri.EscapeDataString(digest)}";
+
+        Uri putUri = x.Uri;
+
+        // TODO: this chunking is super tiny and probably not necessary; what does the docker client do
+        //       and can we be smarter?
+
+        byte[] chunkBackingStore = new byte[MaxChunkSizeBytes];
+
+        int chunkCount = 0;
+        int chunkStart = 0;
+
+        while (contents.Position < contents.Length)
+        {
+            int bytesRead = await contents.ReadAsync(chunkBackingStore);
+
+            ByteArrayContent content = new (chunkBackingStore, offset: 0, count: bytesRead);
+            content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
+            content.Headers.ContentLength = bytesRead;
+            //HttpResponseMessage putResponse = await client.PutAsync(putUri, content);
+
+            //string resp = await putResponse.Content.ReadAsStringAsync();
+
+            //putResponse.EnsureSuccessStatusCode();
+
+            // Chunked upload
+            //content.Headers.ContentRange = new(from:0, to: contents.Length - 1);
+
+            // manual because ACR throws an error with the .NET type {"Range":"bytes 0-84521/*","Reason":"the Content-Range header format is invalid"}
+            //content.Headers.Add("Content-Range", $"0-{contents.Length - 1}");
+            Debug.Assert(content.Headers.TryAddWithoutValidation("Content-Range", $"{chunkStart}-{chunkStart + bytesRead - 1}"));
+
+            HttpResponseMessage? patchResponse = null;
+            for (int retry = 0; retry < 3; retry++)
+            {
+                ExceptionDispatchInfo? e = null;
+                try
+                {
+                    patchResponse = await client.PatchAsync(patchUri, content);
+                    break;
+                }
+                catch (Exception ex)
+                {
+                    e = ExceptionDispatchInfo.Capture(ex);
+                }
+
+                e.Throw();
+            }
```

Comment:
I feel like I'm missing something about this loop - if a particular call fails, we'll capture the exception only to throw it a few lines later? how does this do the additional calls in the retry loop?

Reply:
I don't think it's you that's missing something here! Will simplify.